lib/awful/popup.lua

```diff
@@ -0,0 +1,434 @@
+-- `wibox.layout.flex` is used, then the size would be unlimited and an error
+-- will be printed. The `wibox.layout.fixed`, `wibox.container.constraint`,
+-- `forced_width` and `forced_height` are recommended.
+--
+--@DOC_awful_popup_simple_EXAMPLE@
+--
+-- Here is an example of how to create an alt-tab like dialog by leveraging
+-- the `awful.widget.tasklist`.
+--
+--@DOC_awful_popup_alttab_EXAMPLE@
+--
+-- @author Emmanuel Lepage Vallee
+-- @copyright 2016 Emmanuel Lepage Vallee
+-- @classmod awful.popup
+---------------------------------------------------------------------------
+local wibox     = require( "wibox"           )
+local util      = require( "awful.util"      )
+local glib      = require( "lgi"             ).GLib
+local placement = require( "awful.placement" )
+
+local module = {}
+
+local main_widget = {}
+
+-- Get the optimal direction for the wibox
+-- This (try to) avoid going offscreen
+local function set_position(self)
+    local pf = self._private.placement
+
+    if pf == false then return end
+
+    if pf then
+        pf(self, {bounding_rect = self.screen.geometry})
+        return
+    end
+
+    local geo = self._private.widget_geo
+
+    if not geo then return end
+
+    local preferred_positions = self._private.preferred_directions or {
+        "right", "left", "top", "bottom"
+    }
+
+    local preferred_anchors = self._private.preferred_anchors or {
+        "back", "front", "middle"
+    }
+
+    local _, pos_name, anchor_name = placement.next_to(self, {
+        preferred_positions = preferred_positions,
+        geometry            = geo,
+        preferred_anchors   = preferred_anchors,
+        offset              = self._private.offset or { x = 0, y = 0},
+    })
+
+    if pos_name ~= self._private.current_position then
+        self:emit_signal("property::current_position", pos_name)
+        self._private.current_position = pos_name
+    end
+
+    if anchor_name ~= self._private.current_anchor then
+        self:emit_signal("property::current_anchor", anchor_name)
+        self._private.current_anchor = anchor_name
+    end
+end
+
+-- Fit this widget into the given area
+function main_widget:fit(context, width, height)
+    if not self.widget then
+        return 0, 0
+    end
+
+    return wibox.widget.base.fit_widget(self, context, self.widget, width, height)
+end
+
+-- Layout this widget
+function main_widget:layout(context, width, height)
+    if self.widget then
+        local w, h = wibox.widget.base.fit_widget(self, context, self.widget, 9999, 9999)
+        glib.idle_add(glib.PRIORITY_HIGH_IDLE, function()
+
+            local prev_geo = self._wb:geometry()
+            self._wb.width  = math.max(1, math.ceil(w or 1))
+            self._wb.height = math.max(1, math.ceil(h or 1))
+
+            if self._wb.width ~= prev_geo.width and self._wb.height ~= prev_geo.height then
+                set_position(self._wb)
+            end
+        end)
```

Comment:
> Why does it need to reach for idle_add instead of gears.timer.delayed_call?

The code is very old, I updated it.

> Why does it need to be delayed at all and cannot be done immediately?

An image is worth a thousand word (and you might actually not believe me)

![screenshot_20180115_021433](https://user-images.githubusercontent.com/340384/34930889-d2ae9616-f999-11e7-83bd-baff51eaee4e.png)

My theory is that the xcb surface is not reallocated yet

Reply:
Uhm... Do you happen to know an easy way to reproduce this? At least the following does not work.
```lua
gears.timer.start_new(2, function()
	local w = wibox{ x = 50, y = 50, width = 100, height = 10, bg = "#ff0000", fg = "#000000" }
	w:set_widget(wibox.widget.textbox("Foo"))
	w.visible = true
	gears.timer.delayed_call(function()
		w.height = 50
	end)
end)
```

Reply:
If you apply the patches, it is quite easy to reproduce by commenting the `delayed_call` line. I am sorry for not having a smaller test case. I know about this issue for a years, but I always considered it irrelevant given the kind of hack you have to use to cause it (the popup is such an hack, Radical was another).

Reply:
Ah. Stderr says (however, I did not get visual corruption, only this)
```
2018-01-15 10:03:12 E: awesome: Error during a protected call: lib/wibox/drawable.lua:165: Cairo context entered error state: SURFACE_FINISHED
stack traceback:
	[C]: in function 'assert'
	lib/wibox/drawable.lua:165: in upvalue 'do_redraw'
	lib/wibox/drawable.lua:390: in function <lib/wibox/drawable.lua:388>
	[C]: in function 'xpcall'
	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>
	(...tail calls...)
	lib/gears/timer.lua:231: in function <lib/gears/timer.lua:229>
```
This is the opposite of your theory: The problem is that the xcb surface is reallocated while Lua code still wants to draw to it. Thus, at the end after all the drawing it noticed that something is wrong. (The above happens just by doing `awful.popup{ widget = wibox.widget.textbox("textbox") }`).

So, basically: You may not resize the wibox that is currently drawn while it is being drawn.

...which brings us back to this ugly delayed call and all of its problems.

How much would you be against the popup creating its own `wibox.hierarchy` instance and using that for catching relayouts instead of this magic in `:layout`? That way, this would be disconnected from the widget drawing code instead of being a side effect of it. (And then both your `:fit` and `:layout` are boilerplate and using the argument to `wibox.widget.base.make_widget` makes even more sense... well, no, your whole `main_widget` becomes unnecessary and instead you can just directly give the widget to the wibox)

I wanted to hack this idea into your code, but got lost, so instead here is it in self-contained:
```lua
local widget = wibox.widget.textbox("textbox")
-- Resize the textbox in 5 seconds
gears.timer.start_new(5, function() widget:set_text("This is a longer text than what is was before") end)

local set_widget
do
        local wrapper = wibox.widget.base.make_widget() -- This only needs :layout(), other functions are never used
        local widget = widget
        function wrapper:layout(context, width, height)
                return { wibox.widget.base.place_widget_at(widget, 0, 0,
                        wibox.widget.base.fit_widget(self, context, widget, width, height)) }
        end
        local w = wibox{ x = 50, y = 50, visible = true, widget = widget } -- This displays widget directly, not wrapper!
        local ctxt = { meh_need_to_fake_a_context = true, dpi = screen.primary.dpi }
        local h
        local function layout_callback()
            h:update(ctxt, wrapper, 2^16-1, 2^16-1)
            print(h:get_children()[1]:get_size())
            w.width, w.height = h:get_children()[1]:get_size()
        end
        h = wibox.hierarchy.new(ctxt, wrapper, 2^16-1, 2^16-1, function() end, layout_callback, {})
        layout_callback()
        function set_widget(arg)
                w.widget = arg
                widget = arg
                wrapper:emit_signal("widget::layout_changed")
        end
end

gears.timer.start_new(10, function() set_widget(wibox.widget.textbox("Show a different widget")) end)
```

Reply:
Or alternatively with `:layout` resizing the wibox again directly:
```lua
local widget = wibox.widget.textbox("textbox")
-- Resize the textbox in 5 seconds
gears.timer.start_new(5, function() widget:set_text("This is a longer text than what is was before") end)

local set_widget
do
        local wrapper = wibox.widget.base.make_widget() -- This only needs :layout(), other functions are never used
        local widget = widget
        local w = wibox{ x = 50, y = 50, visible = true, widget = widget } -- This displays widget directly, not wrapper!
        function wrapper:layout(context, width, height)
                local wi, h = wibox.widget.base.fit_widget(self, context, widget, width, height)
                w.width = wi
                w.height = h
                return { wibox.widget.base.place_widget_at(widget, 0, 0, wi, h) }
        end
        local ctxt = { meh_need_to_fake_a_context = true, dpi = screen.primary.dpi }
        local h
        local function layout_callback()
            h:update(ctxt, wrapper, 2^16-1, 2^16-1)
        end
        h = wibox.hierarchy.new(ctxt, wrapper, 0, 0, function() end, layout_callback, {})
        layout_callback()
        function set_widget(arg)
                w.widget = arg
                widget = arg
                wrapper:emit_signal("widget::layout_changed")
        end
end

gears.timer.start_new(10, function() set_widget(wibox.widget.textbox("Show a different widget")) end)
```

Reply:
Umm, I kind of like the idea, but my limited attempts created a result more problematic than the current version. I guess "I got lost" the same way you did when you tried to implement this.